patch/llpcPatchDescriptorLoad.cpp

```diff
@@ -78,6 +78,84 @@ PatchDescriptorLoad::PatchDescriptorLoad()
     initializePatchDescriptorLoadPass(*PassRegistry::getPassRegistry());
 }
 
+// =====================================================================================================================
+/// Create a non-strided, non-swizzled buffer descriptor.
+///
+/// @param baseAddr (i64) base address of the descriptor (top 16 bits are ignored);
+///                 for convenience, an <N x i32> is also accepted (using the low 48 bits)
+/// @param size (i32) size of the referenced buffer in bytes
+Value* PatchDescriptorLoad::buildBufferDescriptor(Value* baseAddr,
+                                                  Value* size,
+                                                  Instruction* insertPt)
+{
+    IRBuilder<> builder(insertPt);
+    Value* bufDesc = UndefValue::get(m_pContext->Int32x4Ty());
+
+    Value* elem0;
+    Value* elem1;
+
+    if (baseAddr->getType() == m_pContext->Int64Ty()) {
+        elem0 = builder.CreateTrunc(baseAddr, m_pContext->Int32Ty());
+        elem1 = builder.CreateLShr(baseAddr, 32, "", true);
+        elem1 = builder.CreateTrunc(elem1, m_pContext->Int32Ty());
+    } else {
+        elem0 = builder.CreateExtractElement(baseAddr, (uint64_t)0);
+        elem1 = builder.CreateExtractElement(baseAddr, (uint64_t)1);
+    }
+
+    // DWORD0
+    bufDesc = builder.CreateInsertElement(bufDesc, elem0, (uint64_t)0);
+
+    // DWORD1
+    elem1 = builder.CreateAnd(elem1, builder.getInt32(UINT16_MAX));
+    bufDesc = builder.CreateInsertElement(bufDesc, elem1, (uint64_t)1);
+
+    // DWORD2
+    bufDesc = builder.CreateInsertElement(bufDesc, size, (uint64_t)2);
+
+    // DWORD3
+#if LLPC_BUILD_GFX10
+    const GfxIpVersion gfxIp = m_pContext->GetGfxIpVersion();
+    if (gfxIp.major < 10)
+#endif
+    {
+        SqBufRsrcWord3 sqBufRsrcWord3 = {};
+        sqBufRsrcWord3.bits.DST_SEL_X = BUF_DST_SEL_X;
+        sqBufRsrcWord3.bits.DST_SEL_Y = BUF_DST_SEL_Y;
+        sqBufRsrcWord3.bits.DST_SEL_Z = BUF_DST_SEL_Z;
+        sqBufRsrcWord3.bits.DST_SEL_W = BUF_DST_SEL_W;
+        sqBufRsrcWord3.gfx6.NUM_FORMAT = BUF_NUM_FORMAT_UINT;
+        sqBufRsrcWord3.gfx6.DATA_FORMAT = BUF_DATA_FORMAT_32;
+        LLPC_ASSERT(sqBufRsrcWord3.u32All == 0x24FAC);
+
+        bufDesc = builder.CreateInsertElement(bufDesc, builder.getInt32(sqBufRsrcWord3.u32All),
+                                              (uint64_t)3);
```

Comment:
120 character line, so this could stay in one line
pBufDesc = builder.CreateInsertElement(...);